From 1e97d3ed94a7968f2dc8266ce5d41242d2bf389c Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Pavel=20Ondra=C4=8Dka?= <pavel.ondracka@gmail.com>
Date: Mon, 13 Apr 2026 22:42:45 +0200
Subject: [PATCH] dri3: add big-endian 8888 fourccs to dri3_cpp_for_fourcc
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Otherwise dri3_alloc_render_buffer() fails on big-endian hosts because
BGRA/BGRX/RGBA/RGBX8888 return cpp=0.

Reviewed-by: Marek Olšák <maraeo@gmail.com>
Part-of: <https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/41132>
---
 src/gallium/frontends/dri/loader_dri3_helper.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/src/gallium/frontends/dri/loader_dri3_helper.c b/src/gallium/frontends/dri/loader_dri3_helper.c
index 8898bd23fdf3..9b1503537dfc 100644
--- a/src/gallium/frontends/dri/loader_dri3_helper.c
+++ b/src/gallium/frontends/dri/loader_dri3_helper.c
@@ -1317,6 +1317,10 @@ dri3_cpp_for_fourcc(uint32_t format) {
    case DRM_FORMAT_ARGB8888:
    case DRM_FORMAT_ABGR8888:
    case DRM_FORMAT_XBGR8888:
+   case DRM_FORMAT_BGRX8888:
+   case DRM_FORMAT_BGRA8888:
+   case DRM_FORMAT_RGBX8888:
+   case DRM_FORMAT_RGBA8888:
    case DRM_FORMAT_XRGB2101010:
    case DRM_FORMAT_ARGB2101010:
    case DRM_FORMAT_XBGR2101010:
-- 
2.43.0

From 8f56c51d51af0af05441bf065529aac55a5227e0 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Pavel=20Ondra=C4=8Dka?= <pavel.ondracka@gmail.com>
Date: Mon, 13 Apr 2026 22:42:45 +0200
Subject: [PATCH] dri: add big-endian 8888 entries to dri2_format_table
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

So that dri2_get_mapping_by_fourcc() resolves the byte-reversed fourccs
(DRM_FORMAT_BGRA/BGRX/RGBA/RGBX8888) used for the native 8888 visual
on big-endian hosts.

Reviewed-by: Marek Olšák <maraeo@gmail.com>
Part-of: <https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/41132>
---
 src/gallium/frontends/dri/dri_helpers.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/src/gallium/frontends/dri/dri_helpers.c b/src/gallium/frontends/dri/dri_helpers.c
index 9cf6d73315ab..68e9ae6395a8 100644
--- a/src/gallium/frontends/dri/dri_helpers.c
+++ b/src/gallium/frontends/dri/dri_helpers.c
@@ -427,16 +427,24 @@ static const struct dri2_format_mapping dri2_format_table[] = {
         PIPE_FORMAT_R10G10B10X2_UNORM, 1, },
       { DRM_FORMAT_ARGB8888,      __DRI_IMAGE_FORMAT_ARGB8888,
         PIPE_FORMAT_BGRA8888_UNORM, 1, },
+      { DRM_FORMAT_BGRA8888,      PIPE_FORMAT_A8R8G8B8_UNORM,
+        PIPE_FORMAT_A8R8G8B8_UNORM, 1, },
       { DRM_FORMAT_ABGR8888,      __DRI_IMAGE_FORMAT_ABGR8888,
         PIPE_FORMAT_RGBA8888_UNORM, 1, },
+      { DRM_FORMAT_RGBA8888,      PIPE_FORMAT_A8B8G8R8_UNORM,
+        PIPE_FORMAT_A8B8G8R8_UNORM, 1, },
       { __DRI_IMAGE_FOURCC_SARGB8888,     __DRI_IMAGE_FORMAT_SARGB8,
         PIPE_FORMAT_BGRA8888_SRGB, 1, },
       { DRM_FORMAT_XRGB8888,      __DRI_IMAGE_FORMAT_XRGB8888,
         PIPE_FORMAT_BGRX8888_UNORM, 1, },
+      { DRM_FORMAT_BGRX8888,      PIPE_FORMAT_X8R8G8B8_UNORM,
+        PIPE_FORMAT_X8R8G8B8_UNORM, 1, },
       { DRM_FORMAT_RGB888,        __DRI_IMAGE_FORMAT_RGB888,
         PIPE_FORMAT_B8G8R8_UNORM, 1, },
       { DRM_FORMAT_XBGR8888,      __DRI_IMAGE_FORMAT_XBGR8888,
         PIPE_FORMAT_RGBX8888_UNORM, 1, },
+      { DRM_FORMAT_RGBX8888,      PIPE_FORMAT_X8B8G8R8_UNORM,
+        PIPE_FORMAT_X8B8G8R8_UNORM, 1, },
       { DRM_FORMAT_BGR888,        __DRI_IMAGE_FORMAT_BGR888,
         PIPE_FORMAT_R8G8B8_UNORM, 1, },
       { DRM_FORMAT_ARGB1555,      __DRI_IMAGE_FORMAT_ARGB1555,
-- 
2.43.0

From c1f1b704d92af4a2de5515fd129b11e0d113eec3 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Pavel=20Ondra=C4=8Dka?= <pavel.ondracka@gmail.com>
Date: Mon, 13 Apr 2026 22:42:46 +0200
Subject: [PATCH] dri: add big-endian 8888 entries to
 driImageFormatToSizedInternalGLFormat
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

So that dma-buf-imported EGLImages on big-endian hosts resolve to a
sized GL internal format in st_bind_egl_image() instead of falling
back to unsized GL_RGBA/GL_RGB.

Reviewed-by: Marek Olšák <maraeo@gmail.com>
Part-of: <https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/41132>
---
 src/gallium/frontends/dri/dri_util.c | 18 ++++++++++++++++++
 1 file changed, 18 insertions(+)

diff --git a/src/gallium/frontends/dri/dri_util.c b/src/gallium/frontends/dri/dri_util.c
index 21e2a7436d82..3b5c22af40d8 100644
--- a/src/gallium/frontends/dri/dri_util.c
+++ b/src/gallium/frontends/dri/dri_util.c
@@ -956,6 +956,24 @@ static const struct {
       .image_format    = __DRI_IMAGE_FORMAT_XBGR8888,
       .internal_format =        GL_RGB8,
    },
+#if UTIL_ARCH_BIG_ENDIAN
+   {
+      .image_format    = PIPE_FORMAT_A8R8G8B8_UNORM,
+      .internal_format =        GL_RGBA8,
+   },
+   {
+      .image_format    = PIPE_FORMAT_X8R8G8B8_UNORM,
+      .internal_format =        GL_RGB8,
+   },
+   {
+      .image_format    = PIPE_FORMAT_A8B8G8R8_UNORM,
+      .internal_format =        GL_RGBA8,
+   },
+   {
+      .image_format    = PIPE_FORMAT_X8B8G8R8_UNORM,
+      .internal_format =        GL_RGB8,
+   },
+#endif
    {
       .image_format    = __DRI_IMAGE_FORMAT_R8,
       .internal_format =        GL_R8,
-- 
2.43.0

